improver/lightning.py

```diff
@@ -204,3 +204,220 @@ def latitude_to_threshold(
+        if np.diff(apcp_time.bound) not in [timedelta(hours=3)]:
+            raise ValueError(
+                f"Precipitation_accumulation cube time window must be three hours, "
+                f"not {np.diff(apcp_time.bound)}."
+            )
+        # Following time/space checks should be made for all diagnostics
+        if cape.coord("forecast_reference_time") != apcp.coord(
+            "forecast_reference_time"
+        ):
+            raise ValueError(
+                "Supplied cubes must have the same forecast reference times"
+            )
+        if not spatial_coords_match([cape, apcp]):
+            raise ValueError("Supplied cubes do not have the same spatial coordinates")
+
+        return cape, liftidx, pwat, cin, apcp
+
+    def process(self, cubes: CubeList, model_id_attr: str = None) -> Cube:
+        """
+        From the supplied CAPE, LIFTIDX, PWAT, CIN, APCP cubes, calculate a probability
+        of lightning cube.
+
+        Args:
+            cubes:
+                Cubes of CAPE, LIFTIDX, PWAT, CIN, APCP.
+            model_id_attr:
+                The name of the dataset attribute to be used to identify the source
+                model when blending data from different models.
+
+        Returns:
+            Cube of lightning data
+
+        Raises:
+            ValueError:
+                If one of the cubes is not found or doesn't match the other
+        """
+        cape, liftidx, pwat, cin, apcp = self._get_inputs(cubes)
+
+        # Regression equations require math on cubes with incompatible units, so strip data
+        templ = apcp.copy()
+        cape = cape.data
+        liftidx = liftidx.data
+        pwat = pwat.data
+        cin = cin.data
+        apcp = apcp.data / 1000 * 39.3701  # convert (kg m-3) to (inches)
+
+        # Regression equation when CAPE and APCP are greater than zero:
+        lprob = cape * apcp
+        lprob = 0.13 * np.log(lprob + 0.7) + 0.05
+
+        # If APCP is very low, a separate regression equation is used to predict lightning
+        # probability. The definition of “very low” is raised slightly when PWAT values are
+        # high. This is because the model often produces showery precipitation that doesn’t
+        # access the actual instability in very moist environments:
+        lprob_noprecip = cape / (cin + 100.0)
+        lprob_noprecip = 0.025 * np.log(lprob_noprecip + 0.31) + 0.03
+        apcp = apcp - (pwat / 1000)
+
+        lprob[np.where(apcp < 0.01)] = lprob_noprecip[np.where(apcp < 0.01)]
+
+        # If there is no CAPE but the atmosphere is “close” to unstable, lightning does sometimes
+        # occur, especially when heavy precipitation may have stabilized the atmosphere in the
+        # model. Unstable values of lifted index are positive here:
+        liftidx = liftidx + 4.0
+
+        liftidx[liftidx < 0] = 0
+
+        lprob[np.where(cape <= 0)] = (
+            0.2 * (liftidx[np.where(cape <= 0)] * apcp[np.where(cape <= 0)]) ** 0.5
```

Comment:
With regard to the two points:

1. You're correct in that `acpc` was modified earlier, and we realized that it was an error in our code as it should not have carried through to this equation.  We have corrected this as we did find it led to some errors that you mentioned in certain circumstances. Here's the snippet of code showing how we changed it so that we created a temporary copy of acpc where it is modified:
```
        apcp_temp = apcp - (pwat / 1000)

        lprob[np.where(apcp_temp < 0.01)] = lprob_noprecip[np.where(apcp_temp < 0.01)]

        # If there is no CAPE but the atmosphere is “close” to unstable, lightning does sometimes
        # occur, especially when heavy precipitation may have stabilized the atmosphere in the
        # model. Unstable values of lifted index are positive here:
        liftidx = liftidx + 4.0

        liftidx[liftidx < 0] = 0

        lprob[np.where(cape <= 0)] = (
            0.2 * (liftidx[np.where(cape <= 0)] * apcp[np.where(cape <= 0)]) ** 0.5
        )
```

2. I agree that this is not an intuitive result, but we've double checked the USAF regression algorithms and the documentation we have and this is in line with the statistics from the analysis that they have conducted.

Reply:
1. Indeed, creating the `apcp_temp` temporary array solves this neatly.

2. Thanks for double-checking. It is curious but I agree about faithfully implementing the approach the USAF developed here